UIEH-371 UIEH-374 Root proxy refactor repository pattern

elrickvm opened the pull request:
## Purpose
The knowledge has the concept of Root Proxy that a user can set at the base level of the eholdings application in their Settings. The goal is to have a dropdown menu that will allow a user to select from all the available Proxy and then send request to update their Root Proxy from the available options in the list. 

Prior we were handling both requests: a) GET - to get all the the available Proxies and b) PUT - to update the root Proxy in one endpoint. That presented a challenge because we were getting a collection of available proxies but needed to return a request of just one proxy and not the collection. This is deemed a Singular Resource.

Separating the GET and PUT into two endpoints will allow the retrieval of the collection and the update of the singular resource to happen independently.

## Approach
  - Separate the concept of Proxy into two entities:
    - `Proxy-Types`
    - `Root-Proxy`
  - Create two endpoint for either entity:
    - `eholdings/proxy-types` 
         _availble requests_
      -  GET - will return all available proxies
    - `eholdings/root-proxy` 
         _availble requests_
      -  GET - will return currently set root-proxy
      -  PUT - will update the root-proxy

## Screenshots
_Let's see those sweet visuals!_
  #### GET all proxy-types
  
<img width="1723" alt="screen shot 2018-05-17 at 9 33 07 pm" src="https://user-images.githubusercontent.com/1953098/40211691-0452963c-5a1a-11e8-9c9f-6625c1bb8e40.png">

  ##### Setting a new root proxy
<img width="1723" alt="screen shot 2018-05-17 at 9 36 16 pm" src="https://user-images.githubusercontent.com/1953098/40211750-60f24144-5a1a-11e8-8abf-df32bc3aa968.png">


  #### Error attempting to set a Root Proxy that is not in the available proxy-types list
<img width="1740" alt="screen shot 2018-05-24 at 8 00 38 pm" src="https://user-images.githubusercontent.com/1953098/40519380-398910b0-5f8d-11e8-99c6-346cb448597b.png">


 #### Error attempting to set a Root Proxy with empty attributes
<img width="1637" alt="screen shot 2018-05-24 at 7 58 57 pm" src="https://user-images.githubusercontent.com/1953098/40519374-310b9d2c-5f8d-11e8-8c83-59e7856f6044.png">

elrickvm committed:
Implement proxy types endpoint

This endpoint will allow the retrieval of a list
of available proxies from the knowledge base.

elrickvm committed:
Implement Root Proxy endpoint

This endpoint will allow for the retrieval of the currently
set root proxy in addition to setting a new root proxy.

sduvvuri-ebsco commented on app/models/proxy_type.rb:
```diff
@@ -0,0 +1,9 @@
+
+  attribute :id
+  attribute :name
+  attribute :url_mask
```
I see that we are not providing "selected" as an attribute here...is the idea that UI will make 2 calls, 1. to get the list and 2. to get the current root-proxy selection to be able to highlight the current selection in the drop-down?

elrickvm replied:
That was my impression from conversation with @sadatay and @cowboyd. I can go either way.

sduvvuri-ebsco commented on app/repositories/root_proxies_repository.rb:
```diff
@@ -0,0 +1,111 @@
+
+      payload[:proxy] = { id: payload.delete(:proxyTypeId) }
+
+      rmapi(:put, '', json: payload)
```
Per our discussion regarding packages refactor, we should add some error handling around the PUT such that if it fails for whatever reason, following GET should not be called?

elrickvm replied:
You are absolutely correct! on the error handling on PUT's. I was waiting to see the approach that was taking so i can copy that pattern in there.

sduvvuri-ebsco commented on app/serializable/serializable_proxy_type.rb:
```diff
@@ -0,0 +1,10 @@
+  # Custom Label attributes
+  attributes :id,
+             :name,
+             :url_mask
```
"" as above regarding "selected"

sduvvuri-ebsco commented on app/repositories/root_proxies_repository.rb:
```diff
@@ -0,0 +1,111 @@
+
+      payload = attrs.to_hash.deep_symbolize_keys
+
+      payload[:proxy] = { id: payload.delete(:proxyTypeId) }
```
Here payload that we get from UI-eholdings will contain only proxy-related information, however, for RM API, proxies and custom-labels are handled by the same PUT call, so we need to make a GET, extract custom-labels and append it to payload before we pass it to RM API; else it would reset all custom-labels.

elrickvm replied:
Yep! We learned that lesson on Thursday evening. I will add the call to get the custom labels today.

sadatay commented on app/controllers/proxy_types_controller.rb:
```diff
@@ -0,0 +1,12 @@
+
+class ProxyTypesController < ApplicationController
+  def index
+    @result = proxy_types.where!
```
it'd be cool to have a method on the repo interface called `all!` that just calls `where!` without params.  just an alias really, but it'd be more readable in situations like this.  definitely not in scope here, just a note for the future

cowboyd replied:
Let's do it! This makes total sense.

sadatay commented on app/controllers/root_proxies_controller.rb:
```diff
@@ -1,43 +1,34 @@
+                                       class: DeserializableRootProxy
+
+  def show
+    @result = root_proxy.find!
```
doesn't this need an id?

elrickvm replied:
Not sure. Because i think it just calls what is deemed the base_url and peels off the proxy

cowboyd replied:
No, because this is a singular resource

cowboyd replied:
For the record, no. because this is a singular resource.

sadatay commented on app/repositories/packages_repository.rb:
```diff
@@ -17,7 +17,6 @@ def initialize(config:)
   # CRUD Operations
   def find!(id)
     request do
-      # binding.pry
```
✂️

elrickvm replied:
100%

sadatay commented on app/repositories/proxy_types_repository.rb:
```diff
@@ -0,0 +1,83 @@
+      'Content-Type': 'application/json',
+      'Accept': 'application/json'
+    }
+  end
```
so, now that there are multiple repositories, there are a bunch of things that we should pull out and centralize with respect to all repo classes.  i'm not saying we should do all of that here, but if there's time maybe we could do something easy like yank these headers out into a superclass?

sadatay commented on app/repositories/proxy_types_repository.rb:
```diff
@@ -0,0 +1,83 @@
+  private
+
+  # superclass for repository errors
+  class RepositoryError < StandardError; end
```
these also would be easy - maybe even easier than the stuff in `initialize`.  they're class definitions, we can just put them in a separate file (or files) in the same namespace and they should be usable by all repos.  no inheritance or anything, just literal copy-paste

sadatay commented on spec/requests/vendors_spec.rb:
```diff
@@ -73,7 +73,6 @@
     let!(:json) { Map JSON.parse response.body }
 
     it 'gets associated package records' do
-      # binding.pry
```
✂️

elrickvm committed:
Use fetch over require in strong params

In cases where a json payload it sent with empty attributes
using require will cause a failure within Ruby on
the ActionParams and bail out. What we rather want is for the
Strong Params to try to get the value and if not found set the
default value which will then continue on into the rest of the code
and hit the Validation block and return a appropriate error
and error code.

elrickvm committed:
Pull custom labels before sending proxy update

If you send a proxy update without sending any preexisting custom labels
that were set. All the custom-labels will be erased.

cowboyd commented on app/controllers/root_proxies_controller.rb:
```diff
@@ -1,43 +1,43 @@
-      Validation::RootProxyParameters.new(data_attributes,
-                                          root_proxy_id,
-                                          root_proxies.all)
+    proxy_types_list = ProxyTypesRepository.new(config: config).all!
```
All of this work is the responsibility of the repository.

the controller method should look like

```ruby
def update
  @result = root_proxy.update! root_proxy_update_params
  render jsonapi: :no_content # 204 No Content or however it's done in rails
end
```

everything else, validation, error handling, pre-requisite reads, are all the responsibility of the repository. The controller is only concerned with HTTP.

elrickvm replied:
@sadatay said that it was settle upon earlier that because of the side-effects that RMAPI does that we would return a body.

sadatay replied:
^ this.  if we want to be consistent with the other endpoints anyway.  the reasoning behind using a 204 depends on knowing with certainty that a resource + an update = current state.  we don't have that certainty elsewhere in rmapi so we return resource bodies on update everywhere else.  

wrt what is and isn't the responsibility of the repository, i do feel like we shouldn't be getting proxytypes here.  it appears to be needed for validation, and maybe this is a scenario where validation shouldn't be in the controller.

cowboyd commented on app/repositories/root_proxies_repository.rb:
```diff
@@ -32,7 +32,14 @@ def update!(attrs)
       # so we need to get the custom labels and post those along
       # with the root-proxy
+
+      _, custom_label_body = rmapi(:get, '')
```
Is this a bug? If the request fails, then it will continue processing.

elrickvm committed:
Refactor Exceptions to use new pattern

The exceptions work is reflected in PR #146

elrickvm committed:
Update Module Descriptor
